### PR TITLE
Rework DynamicTable/DynamicTableRegion.get when the table contains a DTR

### LINKED ISSUE #552: ValueError when using table[int] when table has a ragged DynamicTableRegion
## Description

When a `DynamicTable` contains a column that is a ragged `DynamicTableRegion`, then `table[int]` returns a ValueError. A similar error occurs when getitem is called with a slice or np.ndarray.

This is holding up progress on the ndx-icephys-meta migration. https://github.com/oruebel/ndx-icephys-meta/pull/82

## Steps to Reproduce

```python
# set up one table column to refer to multiple rows of another table using a ragged DynamicTableRegion
from hdmf.common import DynamicTable
dt1 = DynamicTable('test', 'desc')
dt1.add_column('col1', 'desc', index=True)
dt1.add_row(col1=['a', 'b'])
dt1.add_row(col1=['c', 'd'])

dt2 = DynamicTable('test', 'desc')
dt2.add_column('col2', 'desc', index=True, table=dt1)
dt2.add_row(col2=[0, 1])
dt2[0]            # ValueError: Shape of passed values is (2, 2), indices imply (1, 2)
dt2[0:1]          # StopIteration error during print

dt3 = DynamicTable('test', 'desc')
dt3.add_column('col3', 'desc', table=dt1)
dt3.add_row(col3=1)
dt3[0]            # ok, returns dataframe with columns ['col3_id', 'col3_col1'] and one row (1, [c, d])
dt3[0:1]          # ok, returns dataframe with columns ['col3_id', 'col3_col1'] and one row (1, [[c, d]])
```

## Environment

    Python Executable: Conda
    Python Version: Python 3.8
    Operating System: Windows
    HDMF Version: dev

## Checklist

- [x] Have you ensured the bug was not already [reported](https://github.com/hdmf-dev/hdmf/issues) ?
- [x] Have you included a brief and descriptive title?
- [x] Have you included a clear description of the problem you are trying to solve?
- [x] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?

### PR DESCRIPTION
**Pull request by rly**
## Motivation

Rework of #556. Fix #552.

When a `DynamicTable dt` contains a `DynamicTableRegion dtr` and  `dt.get()`, `dt.__getitem__()`, `dtr.get()`, or `dtr.__getitem__()` are called, then:
- If `df=True`, `index=True` (default), then a pandas DataFrame will be returned where the entries for the DTR column are the indices of the DTR. 
- If `df=True` and `index=False`, then a pandas DataFrame will be returned where the entries for the DTR column are nested DataFrames (this works for multiple levels of nesting). 
- If `df=False` and `index=True`, a list of lists will be returned where the list corresponding to the DTR column contains the indices of the DTR.
- If `df=False` and `index=False`, an error will be raised. Returning a list of lists where the DTR column contains lists of lists gets very messy in the ragged DTR case, and the nesting removes all column names and structure, so the returned value is difficult to interpret/parse. It is recommended to set `df=True` if you want to return a row with DynamicTableRegions resolved into nested DataFrames.
- Adds lots of tests to make sure this works for the various combinations of: [`__getitem__`, `get(df=True, index=True)`, `get(df=True, index=False)`, `get(df=False, index=True)`] with the data stored as a [`np.ndarray`, `list`, `h5py.Dataset`] with tables with scalar, 1D arrays, ragged 1D arrays, 2D arrays, DTRs, ragged DTRs at multiple levels of nesting.

Removed name mangling when a `DynamicTable dt` contains a `DynamicTableRegion dtr` and `dt.get()`, `dt.__getitem__()`, `dtr.get()`, or `dtr.__getitem__()` are called. This is technically a breaking change, though it did not work correctly in many cases.

When a `DynamicTable dt` contains a `DynamicTableRegion dtr`, and `dt.to_dataframe()` is called, then:
- If `index=True`, then a pandas DataFrame will be returned where the entries for the DTR column are the indices of the DTR.  This is equivalent to `table.get(slice(None, None, None))`.
- If `index=False` (default), then a pandas DataFrame will be returned where the entries for the DTR column are nested DataFrames (this works for multiple levels of nesting). This is equivalent to `table.get(slice(None, None, None), index=False)`.

Also increase h5py minimum version from 2.9 to 2.10 because 2.9 does not support `__getitem__` on an h5py.Dataset using a list of indices.

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Comment by rly** [comment]
This PR is nearly ready for review. There are two outstanding issues:
1. string handling for h5py 3 in tests
2. ragged compound types (like in the NWB epochs table) are not being handled correctly

**Comment by rly** [comment]
This is now ready for review, tested with h5py 2.10 and 3.3.

**Comment by oruebel** [comment]
It would be useful to add the description of the behavior of the get functions that is included here in the description of the PR also in the tutorial for DynamicTable. https://hdmf.readthedocs.io/en/latest/tutorials/dynamictable.html#sphx-glr-tutorials-dynamictable-py

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,26 +15,23 @@
   `h5py.Dataset` object, and `d[:]` returns `str` objects. Under HDMF 3.x, the same dataset `d` is read
   as a `hdmf.utils.StrDataset` object and `d[:]` still returns `str` objects.
 - Add RRID to docs. @oruebel (#633)
--
+- Allow passing ``index=True`` to ``DynamicTable.to_dataframe()``. @rly (#579)
```
Comment (oruebel) [inline CHANGELOG.md]:
```suggestion
- Allow passing ``index=True`` to ``DynamicTable.to_dataframe()`` to support returning `DynamicTableRegion`` columns as indices or Pandas DataFrame. @rly (#579)
```

**Review by oruebel — APPROVED** [review]

**Review by oruebel — APPROVED** [review]

**Pull request merged**